proof_of_concept/rest/ddm_site.py

```diff
@@ -290,6 +285,97 @@ def on_post(self, request: Request, response: Response) -> None:
             response.body = 'Invalid request'
 
 
+class WorkflowSubmissionHandler:
+    """A handler for the internal /jobs endpoint.
+
+    This lets internal users submit jobs, which are identified by a
+    unique URI.
+
+    """
+    def __init__(self, orchestrator: WorkflowOrchestrator) -> None:
+        """Create a WorkflowSubmissionHandler handler.
+
+        Args:
+            orchestrator: The orchestrator to send requests to.
+
+        """
+        self._orchestrator = orchestrator
+
+    def on_post(self, request: Request, response: Response) -> None:
+        """Handle request to orchestrate a workflow.
+
+        Args:
+            request: The submitted request.
+            response: A response object to configure.
+
+        """
+        if 'requester' not in request.params:
+            logger.info(f'Invalid job submission')
+            response.status = HTTP_400
+            response.body = 'Invalid request'
+            return
+
+        requester = request.params['requester']
+
+        try:
+            site_validator.validate('Job', request.media)
+            job = deserialize(Job, request.media)
+            job_id = self._orchestrator.start_job(requester, job)
+            logger.info(f'Received new job {job_id} from {requester}')
+            response.status = HTTP_303
+            response.location = (
+                    f'{request.forwarded_prefix}/internal/jobs/{job_id}')
+        except ValidationError:
+            logger.warning(f'Invalid execution request: {request.media}')
+            response.status = HTTP_400
+            response.body = 'Invalid request'
```

Comment:
I'd prefer having everything at the end, for two reasons: 1) the code isn't interrupted by error handling all the time (a disadvantage of error return codes), and 2) you have an overview of all the possible results at the end of the function.

I think generally the idea is to handle exceptions in the most logical place, which is here just before ending the function and returning the response to the client. On the other hand, if you have two functions which can return the same type of exception object, and it matters which of the functions raised it, then you'll have to split the try block.

One more note: if doing this, I would add a return to the exception handler rather than using an `else`, because if you have multiple potential issues to handle then your column gets a bit narrower after every one. So an early return is better in that case in my opinion.

Reply:
Interesting, I don't think I have ever encountered someone who prefers this. But I totally agree with your points! Thanks for sharing :)

In my opinion a downside of doing it all at the end is that it is not clear which exceptions can be raised by which parts of the try block. But in this case it is very likely that a method called `validate` might raise this `ValidationError`.

So no need for change here!

Reply:
The thing is that when an error occurs, there are not that many things you can do. Basically, it boils down to 1) try again (for intermittent faults), 2) try something else (if an alternative exists at this level in the call stack, fairly rare), 3) give up and let someone else deal with it (either calling code or, as last resort, the user).

With exceptions, 3) is the default, they bubble up the stack if you don't do anything. In some cases they need to pop out of the top of the stack in a different way than a printed exception trace, so you need to translate. If you have a reactive system processing requests from the outside world, like a WSGI server, then often the error is logger or returned as the result of the request, and then you move on to the next request. In that case, it's important to make sure that no exception can escape and shut down the system (instant denial-of-service attack there). I actually have a bare `except:` clause in Cerise for that. Gets me in trouble with `pylint`.

In this case, the exception is translated into an HTTP status and passed off to the caller that way. WSGI actually turns any escaping exceptions into 500 Internal Server Error (thus avoiding the DOS), but that's not right for an invalid input, so it needs to be handled separately here and turned into the correct code.

In a well-designed library, you have a nice hierarchy of exception types, maybe with machine-readable additional information as needed rather than just an error message, so that you can catch specific ones if needed or catch a broader category where it's all a wash anyway. Thinking about which of the above categories is likely the way a user of the library would handle the exception can help with designing a good set of exception classes. Or you can just be lazy and raise a `RuntimeError` everywhere, like I too often do :-).

I'd like a better solution for 1) actually. The `retrying` package in Python is okay, but it would be nicer to have a `retry(n times)` command you can put into an `except` clause, or `retry(for n seconds)`. Same effect but better syntax, and more granular because it's at the level of a try block, not a function.

Options 1) and 2) are relatively rare, so even if there are many different things that can go wrong, in many different places, often many of these errors need to be handled in the same way simply because there aren't many options. In that case catching them all at once is nice and concise.

With error returns, you end up explicitly handling possible errors for every single call, often just passing them on up the stack explicitly. That does have the advantage of it being very clear what's going on, and some people like it because of that (like Linus Torvalds, one of his reasons for not switching the kernel to C++). The designers of Go also use return types, but their idea is more that most errors aren't actually all that exceptional, and should just be considered possible results of calling a function, and handled in place. I do like that actually, and their way of returning a `result, error` tuple is cleaner than C's `errno`. I suspect it also lets you do things with those mystical monads. Anyway, they do have an exception system in Go, but it's called `panic` and it's intended for truly exceptional things like not being able to allocate memory (quite hard to do on modern systems). I'm not completely against this, but it gets incredibly tedious sometimes...